tap_salesforce/salesforce/bulk.py

```diff
@@ -117,25 +118,28 @@ def _bulk_query(self, catalog_entry, state):
 
         if batch_status['state'] == 'Failed':
             if self._can_pk_chunk_job(batch_status['stateMessage']):
-                batch_status = self._bulk_query_with_pk_chunking(catalog_entry, start_date)
-                job_id = batch_status['job_id']
-
-                # Set pk_chunking to True to indicate that we should write a bookmark differently
-                self.sf.pk_chunking = True
-
-                # Add the bulk Job ID and its batches to the state so it can be resumed if necessary
-                tap_stream_id = catalog_entry['tap_stream_id']
-                state = singer.write_bookmark(state, tap_stream_id, 'JobID', job_id)
-                state = singer.write_bookmark(state, tap_stream_id, 'BatchIDs', batch_status['completed'][:])
-
-                for completed_batch_id in batch_status['completed']:
-                    for result in self.get_batch_results(job_id, completed_batch_id, catalog_entry):
-                        yield result
-                    # Remove the completed batch ID and write state
-                    state['bookmarks'][catalog_entry['tap_stream_id']]["BatchIDs"].remove(completed_batch_id)
-                    LOGGER.info("Finished syncing batch %s. Removing batch from state.", completed_batch_id)
-                    LOGGER.info("Batches to go: %d", len(state['bookmarks'][catalog_entry['tap_stream_id']]["BatchIDs"]))
-                    singer.write_state(state)
+                # Get list of batch_status with pk_chunking or date_windowing
+                status_list = self._bulk_with_window([], catalog_entry, start_date)
+
+                for batch_status in status_list:
+                    job_id = batch_status['job_id']
+
+                    # Set pk_chunking to True to indicate that we should write a bookmark differently
+                    self.sf.pk_chunking = True
+
+                    # Add the bulk Job ID and its batches to the state so it can be resumed if necessary
+                    tap_stream_id = catalog_entry['tap_stream_id']
+                    state = singer.write_bookmark(state, tap_stream_id, 'JobID', job_id)
+                    state = singer.write_bookmark(state, tap_stream_id, 'BatchIDs', batch_status['completed'][:])
+
+                    for completed_batch_id in batch_status['completed']:
+                        for result in self.get_batch_results(job_id, completed_batch_id, catalog_entry):
+                            yield result
+                        # Remove the completed batch ID and write state
+                        state['bookmarks'][catalog_entry['tap_stream_id']]["BatchIDs"].remove(completed_batch_id)
+                        LOGGER.info("Finished syncing batch %s. Removing batch from state.", completed_batch_id)
```

Comment:
```suggestion
                        LOGGER.info("Finished syncing batch %s. Removed batch from state.", completed_batch_id)
```
Since this log comes after the actual `remove()`, we want it to be clear that if you see this line, the batch was removed

Reply:
Updated.